util/git/client.go

```diff
@@ -201,25 +223,76 @@ func (m *nativeGitClient) Checkout(revision string) error {
 	return nil
 }
 
-// LsRemote returns the commit SHA of a specific branch, tag, or HEAD
+// LsRemote resolves the commit SHA of a specific branch, tag, or HEAD. If the supplied revision
+// does not resolve, and "looks" like a 7+ hexadecimal commit SHA, it return the revision string.
+// Otherwise, it returns an error indicating that the revision could not be resolved. This method
+// runs with in-memory storage and is safe to run concurrently, or to be run without a git
+// repository locally cloned.
 func (m *nativeGitClient) LsRemote(revision string) (string, error) {
-	var args []string
-	if revision == "" || revision == "HEAD" {
-		args = []string{"ls-remote", "origin", "HEAD"}
-	} else {
-		args = []string{"ls-remote", "--head", "--tags", "origin", revision}
-
+	if IsCommitSHA(revision) {
+		return revision, nil
+	}
+	repo, err := git.Init(memory.NewStorage(), nil)
+	if err != nil {
+		return "", err
+	}
+	remote, err := repo.CreateRemote(&config.RemoteConfig{
+		Name: git.DefaultRemoteName,
+		URLs: []string{m.repoURL},
+	})
+	if err != nil {
+		return "", err
 	}
-	out, err := m.runCmd("git", args...)
+	refs, err := remote.List(&git.ListOptions{})
 	if err != nil {
 		return "", err
 	}
-	if out == "" {
-		// if doesn't exist in remote, assume revision is a commit sha and return it
+	if revision == "" {
+		revision = "HEAD"
+	}
+	// refToHash keeps a maps of remote refs to their hash
+	// (e.g. refs/heads/master -> a67038ae2e9cb9b9b16423702f98b41e36601001)
+	refToHash := make(map[string]string)
+	// refToResolve remembers ref name of the supplied revision if we determine the revision is a
+	// symbolic reference (like HEAD), in which case we will resolve it from the refToHash map
+	refToResolve := ""
+	for _, ref := range refs {
+		refName := ref.Name().String()
+		if refName != "HEAD" && !strings.HasPrefix(refName, "refs/heads/") && !strings.HasPrefix(refName, "refs/tags/") {
+			// ignore things like 'refs/pull/' 'refs/reviewable'
+			continue
+		}
+		hash := ref.Hash().String()
+		if ref.Type() == plumbing.HashReference {
+			refToHash[refName] = hash
+		}
+		//log.Debugf("%s\t%s", hash, refName)
+		if ref.Name().Short() == revision {
+			if ref.Type() == plumbing.HashReference {
+				log.Debugf("revision '%s' resolved to '%s'", revision, hash)
```

Comment:
@merenbach I actually prefer our messages to use single quotes instead of double quotes, which is why I use `'%s'` and not `%q`.

Reply:
Thanks, @jessesuen!  Have opened https://github.com/argoproj/argo-cd/issues/579 to help us ensure we're on the same page moving forward, since I've been using `%q` and should change to conform.